### PR TITLE
Add the framework for the pg_statistic table.

### PR DESCRIPTION
**Pull request by preetansh**
# Heading

Add the framework for the `pg_statistic` table.

## Description

This PR adds the skeleton of `pg_statistic` support.
It does not currently track any statistics at all, apart from creating dummy column statistics in `pg_statistic` and cleaning those entries up.

### CODE REVIEW

File: src/catalog/postgres/builder.cpp [hunk 1]
```diff
@@ -670,6 +676,31 @@ IndexSchema Builder::GetLanguageNameIndexSchema(db_oid_t db) {
...
+
+  columns.emplace_back("starelid", type::TypeId::INTEGER, false,
+                       parser::ConstantValueExpression(type::TypeId::INTEGER));
+  columns.back().SetOid(PgStatistic::STARELID_COL_OID);
+
+  columns.emplace_back("staattnum", type::TypeId::INTEGER, false,
+                       parser::ConstantValueExpression(type::TypeId::INTEGER));
+  columns.back().SetOid(PgStatistic::STAATTNUM_COL_OID);
+
+//  columns.emplace_back("stanullfrac", type::TypeId::DECIMAL, false,
```
Comment (lmwnshn) [inline src/catalog/postgres/builder.cpp:690]: Please remove dead code.

File: src/catalog/postgres/pg_core_impl.cpp [hunk 2]
```diff
@@ -7,6 +7,7 @@
 #include "catalog/postgres/pg_class.h"
 #include "catalog/postgres/pg_index.h"
 #include "catalog/postgres/pg_namespace.h"
+#include "catalog/postgres/pg_statistic.h"
```
Comment (lmwnshn) [inline src/catalog/postgres/pg_core_impl.cpp:10]: Could you move all the changes from this file out of `pg_core_impl` and into its own `pg_statistic_impl`?

The idea is that `pg_core` is the absolute minimum necessary to spin up the rest of the system.
We spin up the system now just fine without stats, so it doesn't belong in `core`.

File: src/catalog/postgres/pg_core_impl.cpp [hunk 3]
```diff
@@ -1525,6 +1558,119 @@ Column PgCoreImpl::MakeColumn(storage::ProjectedRow *const pr, const storage::Pr
...
 }
 
+template <typename Column, typename ClassOid, typename ColOid>
+void PgCoreImpl::CreateColumnStatistic(const common::ManagedPointer<transaction::TransactionContext> txn,
+                                            const ClassOid class_oid, const ColOid col_oid, const Column &col) {
+  // Step 1: Insert into the table
+  auto *const redo = txn->StageWrite(db_oid_, PgStatistic::STATISTIC_TABLE_OID, pg_statistic_all_cols_pri_);
+
+  // Write the attributes in the Redo Record
+  auto relid_entry = reinterpret_cast<ClassOid *>(
```
Comment (lmwnshn) [inline src/catalog/postgres/pg_core_impl.cpp:1568]: Could you look at how the new code handles accesses of projected rows?

File: src/catalog/postgres/pg_core_impl.cpp [hunk 4]
```diff
@@ -1525,6 +1558,119 @@ Column PgCoreImpl::MakeColumn(storage::ProjectedRow *const pr, const storage::Pr
...
+      redo->Delta()->AccessForceNotNull(pg_statistic_all_cols_prm_[PgStatistic::STARELID_COL_OID]));
+  auto attnum_entry = reinterpret_cast<ColOid *>(
+      redo->Delta()->AccessForceNotNull(pg_statistic_all_cols_prm_[PgStatistic::STAATTNUM_COL_OID]));
+//  auto nullfrac_entry = reinterpret_cast<double *>(
```
Comment (lmwnshn) [inline src/catalog/postgres/pg_core_impl.cpp:1572]: Please remove all dead code. This comment applies throughout the rest of the file and rest of the PR.

File: src/catalog/postgres/pg_core_impl.cpp [hunk 5]
```diff
@@ -522,8 +547,10 @@ bool PgCoreImpl::CreateTableEntry(const common::ManagedPointer<transaction::Tran
   {
     col_oid_t curr_col_oid(1);
     for (auto &col : schema.GetColumns()) {
-      auto success = CreateColumn(txn, table_oid, curr_col_oid++, col);
+      col_oid_t col_oid(curr_col_oid++);
+      auto success = CreateColumn(txn, table_oid, col_oid, col);
       if (!success) return false;
+      CreateColumnStatistic(txn, table_oid, col_oid, col);
```
Comment (lmwnshn) [inline src/catalog/postgres/pg_core_impl.cpp:553]: Must column statistics be created alongside a column?

File: src/include/catalog/database_catalog.h [hunk 6]
```diff
@@ -12,6 +12,7 @@
 #include "catalog/postgres/pg_proc_impl.h"
 #include "catalog/postgres/pg_type_impl.h"
 #include "common/managed_pointer.h"
+#include "optimizer/statistics/table_stats.h"
```
Comment (lmwnshn) [inline src/include/catalog/database_catalog.h:15]: Unnecessary include.

File: src/include/catalog/postgres/pg_core_impl.h [hunk 7]
```diff
@@ -323,6 +326,32 @@ class PgCoreImpl {
   template <typename Column, typename ColOid>
   static Column MakeColumn(storage::ProjectedRow *pr, const storage::ProjectionMap &pr_map);
 
+  /**
+   * Add entry to pg_statistic
+   *
+   * Currently, this is called inside CreateTableEntry so that each row in pg_attribute (corresponding to a column
```
Comment (lmwnshn) [inline src/include/catalog/postgres/pg_core_impl.h:332]: Couldn't this be done by the caller of `CreateTableEntry` later?

File: src/include/catalog/postgres/pg_statistic.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+ private:
+  friend class Builder;
+  friend class PgCoreImpl;
+
+  /*
+   * Column names of the form "STA[name]_COL_OID" are present in the PostgreSQL
+   * catalog specification and columns of the form "STA_[name]_COL_OID" are
+   * terrier-specific additions (generally pointers to internal objects).
+   */
+  static constexpr col_oid_t STARELID_COL_OID = col_oid_t(1);     // INTEGER (fkey: pg_class) [table_oid_t]
```
Comment (lmwnshn) [inline src/include/catalog/postgres/pg_statistic.h:28]: Could you look at how `pg_proc` or similar does these now?

**Review by lmwnshn — CHANGES_REQUESTED** [review]
very quick review, please change to new-style catalog stuff.

File: src/catalog/postgres/builder.cpp (hunk 1 above)
Comment (jkosh44) [reply src/catalog/postgres/builder.cpp:690]: @whoever is implementing the PR, Is this dead or not yet implemented? I think we'll need these columns as well as some binary array columns for topKs and histograms.

File: src/catalog/postgres/pg_core_impl.cpp (hunk 5 above)
Comment (arvindsaik) [reply src/catalog/postgres/pg_core_impl.cpp:553]: We did not know where else to create it. Do you have any suggestions?

File: src/catalog/postgres/pg_core_impl.cpp (hunk 4 above)
Comment (arvindsaik) [reply src/catalog/postgres/pg_core_impl.cpp:1572]: Done

File: src/include/catalog/database_catalog.h (hunk 6 above)
Comment (arvindsaik) [reply src/include/catalog/database_catalog.h:15]: Removed

File: src/include/catalog/postgres/pg_statistic.h (hunk 8 above)
Comment (arvindsaik) [reply src/include/catalog/postgres/pg_statistic.h:28]: Done

File: src/catalog/postgres/pg_core_impl.cpp (hunk 3 above)
Comment (arvindsaik) [reply src/catalog/postgres/pg_core_impl.cpp:1568]: Done

File: src/catalog/postgres/pg_core_impl.cpp (hunk 2 above)
Comment (arvindsaik) [reply src/catalog/postgres/pg_core_impl.cpp]: Done

File: src/include/catalog/postgres/pg_core_impl.h (hunk 7 above)
Comment (arvindsaik) [reply src/include/catalog/postgres/pg_core_impl.h]: Done.

File: src/catalog/postgres/builder.cpp (hunk 1 above)
Comment (lmwnshn) [reply src/catalog/postgres/builder.cpp]: @jkosh44 You'll probably be the one implementing it.

File: src/catalog/postgres/pg_core_impl.cpp (hunk 2 above)
Comment (lmwnshn) [reply src/catalog/postgres/pg_core_impl.cpp]: thanks!

File: src/catalog/postgres/pg_core_impl.cpp (hunk 3 above)
Comment (lmwnshn) [reply src/catalog/postgres/pg_core_impl.cpp]: nice work. thanks!

**Comment by lmwnshn** [comment]
Assuming this passes CI, @jkosh44 can you take a review pass since you'll be working with the PR and I've made some changes so I probably shouldn't be reviewing myself? Thanks!

**Review by jkosh44 — APPROVED** [review]
LGTM, though I'm not super familiar with the Catalog

**Review by lmwnshn — APPROVED** [review]
Rubber stamping.

**Pull request merged**